gbdk-lib/examples/cross-platform/text_typewriter/src/main.c

```diff
@@ -0,0 +1,108 @@
+#include <gbdk/platform.h>
+#include <stdint.h>
+#include "Font.h"
+
+uint8_t GetCharacterVRamTile(char character) {
+
+    uint8_t vramTile=0;
+
+    // Char's can be interpreted as integers
+    // We don't need to map every alpha-numeric character
+    // We can use basic math to simplify A-Z and 0-9
+    if(character>='a'&&character<='z')vramTile = (character-'a')+1;
+    else if(character>='A'&&character<='Z')vramTile = (character-'A')+1;
+    else if(character>='0'&&character<='9')vramTile = (character-'0')+27;
+    else if(character=='!')vramTile = 37;
+    else if(character==':')vramTile = 38;
+    else if(character=='?')vramTile = 39;
+    else if(character=='/')vramTile = 40;
+    else if(character=='=')vramTile = 41;
+    else if(character==',')vramTile = 42;
+    else if(character=='.')vramTile = 43;
+    else if(character=='<')vramTile = 44;
+    else if(character=='>')vramTile = 45;
+
+    return vramTile;
+
+}
+
+
+void DrawText( char* text,uint8_t typewriterDelay){
+
+    uint8_t column = 0;
+    uint8_t row = 0;
+
+    // Get the address of the first tile in the row
+    uint8_t* vramAddress = get_bkg_xy_addr(column,row);
+
+    uint16_t index=0;
+    uint8_t columnIndex=0;
+
+    while(text[index]!='\0'){
+
+        char character = text[index];
+
+        // Get the vram tile to draw
+        uint8_t vramTile = GetCharacterVRamTile(character);
+
+        set_vram_byte(vramAddress++,vramTile);
+
+        #if defined(SEGA)
+        set_vram_byte(vramAddress++,0);
+        #endif
+
+        index++;
+        columnIndex++;
+
+        // if we've reached the end of the row
+        if(columnIndex>=DEVICE_SCREEN_WIDTH){
+            
+            // reset for the next row
+            vramAddress = get_bkg_xy_addr(column,++row);
+            
+            columnIndex=0;
+        }
+
+        if(typewriterDelay>0){
+
+            #if !defined(SEGA) && !defined(NINTENDO_NES)
+            
+            // Play a basic sound effect
+            NR10_REG = 0x34;
+            NR11_REG = 0x81;
+            NR12_REG = 0x41;
```

Comment:
For the MegaDuck values written to NR12 need a Nybble swap (true for other channels as well, but not applicable here). There's an example of this in `examples/gb/sound`

```C
inline uint8_t translate_envelope(uint8_t value) {
	#ifdef MEGADUCK
        // Mega Duck has nybbles swapped for NR12, NR22, NR42 audio envelope registers
		return ((uint8_t)(value << 4) | (uint8_t)(value >> 4));
	#else
		return value;
	#endif
}
```
Then
```C
NR12_REG = translate_envelope(0x41);
```

https://github.com/gbdk-2020/gbdk-2020/blob/c97d278cd4178dbe0bd3d66589821d7f6f31e711/gbdk-lib/examples/gb/sound/sound.c#L7C1-L14C2

MegaDuck sound info in the GBDK docs:
https://gbdk-2020.github.io/gbdk-2020/docs/api/docs_supported_consoles.html#autotoc_md190